### DISCUSSION TITLE
How to get the actual typeof(<JSON Array Element>).name ?

### DISCUSSION THREAD

**Discussion by retoxing** [vlang/v#22817] [Questions and Answers] [answered]
Hi everyone,

`
tx := "[\"ABC\", 123, 4.56]"
js := json2.raw_decode(tx)!
ja := js.arr()
for vl in ja {
    dump(typeof(vl).name)
}
`

The above outputs **json2.Any** as the type for each array element, how do I get the actual data type?

**Comment by retoxing** [comment]
> Hi everyone,
> 
> `tx := "[\"ABC\", 123, 4.56]"js := json2.raw_decode(tx)! ja := js.arr() for vl in ja { dump(typeof(vl).name) }`
> 
> The above outputs **json2.Any** as the type for each array element, how do I get the actual data type?

I don't know why the post removed my line breaks from the code :-(

**Comment by jorgeluismireles** [comment]
I would do this:
```v
import json

type Elem = string | f64

tx := [
	Elem('ABC')
	Elem(f64(123))
	Elem(4.56)
]

assert json.encode(tx).str() == '["ABC",123,4.56]'

```
https://play.vlang.io/p/7f5518d3e4

**Reply by retoxing** [reply]
Hi Jorge,

I want to programmatically determine the data types of elements received in a JSON encoded array, your example just takes me from **json2.Any, json2.Any, json2.Any** to **Elem, Elem, Elem** which is not really progress. 

In the example array I gave I would want the answer to be **string, int, f64**

**Reply by jorgeluismireles** [reply, marked as answer]
Yes, the inverse would be:
```v
import json

type Elem = string | f64

elems := json.decode([]Elem, '["ABC",123,4.56]')!
for elem in elems {
	match elem {
		string {
			println('string: ${elem}')	
		}
		f64 {
			println('float: ${elem}')
		}
	}
}
```
The result of the above is:
```
string: ABC
float: 123.0
float: 4.56
```
And more work is needed to get apart the `ints` from the `floats`.

Imho using a **bag** in json holding an array of diverse objects is sooner or later more difficult too handle.

I'll prefer to have an array with named elements, for instance:
```v
import json

struct Samples {
	ys    []f64
	xs    []int
	units string
}

samples := Samples{
	ys: [ 0.0, 10.3, 20.5 ]
	xs: [ 0, 1, 2 ]
	units: 'ohms/sec'
}

assert json.encode(samples).str() == '{"ys":[0,10.3,20.5],"xs":[0,1,2],"units":"ohms/sec"}'
```
That way, you don't have to check the type of any array's element and later guess what the value meaning is.

A *bag* of json diverse objects seems to be more compact, but requires more code to read and write.

Regards.

**Reply by retoxing** [reply]
Thank you, Jorge.

> And more work is needed to get apart the ints from the floats

Using match{} with json2.Any also identifies the ints
